src/kernel/filesystem/fat32.zig

```diff
@@ -1277,34 +1377,672 @@ pub fn Fat32FS(comptime StreamType: type) type {
+        ///
+        /// Return: u32
+        ///     The next free cluster to use.
+        ///
+        /// Error: Allocator.Error || ReadError || SeekError || Fat32Self.Error
+        ///     Allocator.Error - Not enough memory for allocating memory.
+        ///     WriteError      - Error while updating the FAT with the new cluster.
+        ///     ReadError       - Error while reading the stream.
+        ///     SeekError       - Error while seeking the stream.
+        ///     Fat32Self.Error.BadRead  - Error reading the FAT, not aligned to the sector.
+        ///     Fat32Self.Error.DiskFull - No free clusters.
+        ///
+        fn findNextFreeCluster(self: *Fat32Self, cluster_hint: u32, parent_cluster: ?u32) (Allocator.Error || WriteError || ReadError || SeekError || Fat32Self.Error)!u32 {
+            var fat_buff = try self.allocator.alloc(u32, self.fat_config.bytes_per_sector / @sizeOf(u32));
+            defer self.allocator.free(fat_buff);
+            var sector_offset = cluster_hint / fat_buff.len;
+
+            const reader = self.stream.reader();
+            const writer = self.stream.writer();
+            const seeker = self.stream.seekableStream();
+
+            try seeker.seekTo((self.fat_config.reserved_sectors + sector_offset) * self.fat_config.bytes_per_sector);
+            var fat_read = try reader.readAll(std.mem.sliceAsBytes(fat_buff));
+            if (fat_read != self.fat_config.bytes_per_sector) {
+                return Fat32Self.Error.BadRead;
+            }
+
+            // Check for a free cluster by checking the FAT for a 0x00000000 entry (free)
+            var cluster = cluster_hint;
+            while (fat_buff[cluster - (sector_offset * fat_buff.len)] != 0x00000000) {
+                cluster += 1;
+
+                // Check we are still in the FAT buffer, if not, read the next FAT part
+                const check_offset = cluster / fat_buff.len;
+                if (check_offset > sector_offset) {
+                    // Check if the cluster will go outside the FAT
+                    if (check_offset >= self.fat_config.sectors_per_fat) {
+                        // TODO: Will need to wrap as there maybe free clusters before the hint
+                        if (self.fat_config.has_fs_info) {
+                            std.debug.assert(self.fat_config.number_free_clusters == 0);
+                        }
+                        return Fat32Self.Error.DiskFull;
+                    }
+                    sector_offset = check_offset;
+                    try seeker.seekTo((self.fat_config.reserved_sectors + sector_offset) * self.fat_config.bytes_per_sector);
+                    fat_read = try reader.readAll(std.mem.sliceAsBytes(fat_buff));
+                    if (fat_read != fat_buff.len * @sizeOf(u32)) {
+                        return Fat32Self.Error.BadRead;
+                    }
+                }
+            }
+
+            // Update the FAT for the allocated cluster
```

Comment:
But what if the restore fails as will still need to write to the disk? I think caching might help, but I don't think I can have all the writes happen atomically. Plus, the write will fail if there is an issue with the disk and that would be a bigger problem.

Reply:
Would one of these writes failing cause any problems for other clusters or the rest of the filesystem? If it would only affect this one cluster/file then I don't see it being corrupted as too big an issue as long as the error gets reported properly so the user knows.

Reply:
This write would only effect either a newly created file, writing to a file that is bigger than a cluster. so could corrupt the file being worked on and nothing else (I think). The error will bubble up to who called write or open create. It would result in lost clusters, but won't effect the file system as a whole. But depending on which write failed would corrupt the file system in different ways.